src/contracts/data-bus/data-bus.service.ts

```diff
@@ -0,0 +1,186 @@
+import { Inject, Injectable, LoggerService } from '@nestjs/common';
+import { getToken, InjectMetric } from '@willsoto/nestjs-prometheus';
+import { OneAtTime } from 'common/decorators';
+import {
+  METRIC_ACCOUNT_BALANCE,
+  METRIC_RPC_REQUEST_DURATION,
+  METRIC_RPC_REQUEST_ERRORS,
+} from 'common/prometheus';
+import { WINSTON_MODULE_NEST_PROVIDER } from 'nest-winston';
+import { Counter, Gauge, Histogram, register } from 'prom-client';
+import { RpcProvider } from 'provider';
+import {
+  DATA_BUS_ADDRESS,
+  DATA_BUS_BALANCE_UPDATE_BLOCK_RATE,
+  DATA_BUS_PRIVATE_KEY,
+  DATA_BUS_PROVIDER_CONFIG_PATH,
+} from './data-bus.constants';
+
+import { Configuration } from 'common/config';
+import { DataBusClient } from './data-bus.client';
+import { MessageRequiredFields } from 'messages';
+import { DSMMessageSender } from './dsm-message.sender.client';
+import { getProviderFactory } from 'provider/provider.factory';
+import { StaticJsonRpcProvider } from '@ethersproject/providers';
+import { ModuleRef } from '@nestjs/core';
+
+@Injectable()
+export class DataBusService {
+  private dsmMessageSender!: DSMMessageSender;
+  private provider!: RpcProvider;
+  constructor(
+    @InjectMetric(METRIC_ACCOUNT_BALANCE) private accountBalance: Gauge<string>,
+    @Inject(WINSTON_MODULE_NEST_PROVIDER) private logger: LoggerService,
+    @Inject(DATA_BUS_PRIVATE_KEY) private privateKey: string,
+    @Inject(DATA_BUS_ADDRESS) private dataBusAddress: string,
+    protected readonly config: Configuration,
+    @Inject(getToken(METRIC_RPC_REQUEST_DURATION))
+    private rpcReqDurationMetric: Histogram<string>,
+    @Inject(getToken(METRIC_RPC_REQUEST_ERRORS))
+    private rpcReqErrorsMetric: Counter<string>,
+    private moduleRef: ModuleRef,
+  ) {}
+
+  async initialize() {
+    try {
+      this.provider = await this.createProvider();
+
+      const guardianAddress = this.address;
+      register.setDefaultLabels({ guardianAddress });
+
+      const dataBusClient = new DataBusClient(this.dataBusAddress, this.wallet);
+      this.dsmMessageSender = new DSMMessageSender(dataBusClient);
+      await this.monitorGuardianBalance();
+      this.subscribeToEthereumUpdates();
+    } catch (error) {
+      this.logger.error(error);
+    }
+  }
+
+  /**
+   * Subscribes to the event of a new block appearance
+   */
+  public subscribeToEthereumUpdates() {
+    const provider = this.provider;
+    this.provider.clone();
+    provider.on('block', async (blockNumber) => {
+      if (blockNumber % DATA_BUS_BALANCE_UPDATE_BLOCK_RATE !== 0) return;
+      await this.monitorGuardianBalance().catch((error) =>
+        this.logger.error(error),
+      );
+    });
+
+    this.logger.log('WalletService subscribed to Ethereum events');
+  }
+
+  /**
+   * Monitors the guardian account balance to ensure it is sufficient for transactions.
+   * Updates the account balance metric.
+   */
+  @OneAtTime()
+  public async monitorGuardianBalance() {
```

Comment:
we already have guardian balance monitoring  in WalletService 
Can we use it here? 
we also now set metric value METRIC_ACCOUNT_BALANCE in two places

Reply:
WalletService looks like more proper place for checking balance of account

Reply:
It's a separate network, but with the same wallet

Reply:
got it
but metric should be separate for data bus 
METRIC_ACCOUNT_BALANCE  is already in use  and it is value very important for sending unvet trans

Reply:
Thanks for noticing! I've corrected the metrics and logs.